### PR TITLE
Added SsoConfig to read one time configuration for client apps

### PR DESCRIPTION
**Pull request by lalwani**
Description:
Created SsoConfig object to read one-time config information from an sso_config.json file to be declared in the client app as a raw resource.
There are three configurations to be read - client_id, scope and redirect_uri.

Other configurations like discovery_uri will be added later.

### CODE REVIEW

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import org.json.JSONException
+import org.json.JSONObject
+
+data class SsoConfig(val clientId: String, val redirectUri: String, val scope: String? = null)
+
+object SsoConfigProvider {
+  fun getSsoConfig(context: Context): SsoConfig {
+    val resources: Resources = context.resources
+    val resourceId = resources.getIdentifier(SSO_CONFIG_FILE, "raw", context.packageName)
+    val configSource: BufferedSource =
```
Comment (skhanzada) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt:39]: don't you need to close this?

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      Okio.buffer(Okio.source(resources.openRawResource(resourceId)))
+    val configData = Buffer()
+    try {
+      configSource.readAll(configData)
+      val configJson = JSONObject(configData.readString(Charset.forName("UTF-8")))
+      val clientId = getRequiredConfigString(configJson, CLIENT_ID_PARAM)
+      val scope = getConfigString(configJson, SCOPE_PARAM)
+      val redirectUri = getRequiredConfigString(configJson, REDIRECT_PARAM)
+      configSource.close()
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt:48]: Can we use the `.use { }` extension on `Closeable` instead? 
Resources are a very delicate matter, and leaking them is a real risk, I'd prefer we use battle-tested code like `.use { }` extension (equivalent to Java's `try-with-resources`)

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return getConfigString(configJson, propName)
+      ?: throw AuthException.ClientError(
+        "$propName is required but not specified in the configuration"
+      )
+  }
+
+  private fun getConfigString(configJson: JSONObject, propName: String): String? {
+    var value: String = configJson.optString(propName) ?: return null
+    value = value.trim { it <= ' ' }
+    return if (TextUtils.isEmpty(value)) {
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt:68]: You could just do
```
return value.trim { it <= ' ' }.takeIf { it.isNotEmpty() }
```

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt (hunk 2 above)
Comment (lalwani) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt:48]: I tried that but it requires to pull in a different version of okio then the one retrofit uses right now. And retrofit lib is the latest stable version

Comment (psteiger) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt:48]: > I tried that but it requires to pull in a different version of okio then the one retrofit uses right now. And retrofit lib is the latest stable version

Why so? Isn't `configSource` `Closeable` already?

And `use { }` is from Kotlin stdlib: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

Comment (lalwani) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt]: hmm.. you're right. Maybe I confused it with another dependency. pushed another patch with  `use { }`

**Review by psteiger — APPROVED** [review]

File: authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        configSource.readAll(configData)
+        val configJson = JSONObject(configData.readString(Charset.forName("UTF-8")))
+        val clientId = getRequiredConfigString(configJson, CLIENT_ID_PARAM)
+        val scope = getConfigString(configJson, SCOPE_PARAM)
+        val redirectUri = getRequiredConfigString(configJson, REDIRECT_PARAM)
+        return SsoConfig(clientId, redirectUri, scope)
+      } catch (ex: IOException) {
+        throw AuthException.ClientError("Failed to read configuration: " + ex.message)
+      } catch (ex: JSONException) {
+        throw AuthException.ClientError("Failed to read configuration: " + ex.message)
```
[the remark this answers is not in the document]
Comment (lalwani) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/api/request/SsoConfig.kt:52]: Don't really see much use of it for the client apps so its probably better to hide the details. That's why encapsulated the error inside our provided AuthException

**Pull request merged**